comp_arguments.go

```diff
@@ -0,0 +1,123 @@
+	ShellCompDirectiveError ShellCompDirective = 1 << iota
+
+	// ShellCompDirectiveNoSpace indicates that the shell should not add a space
+	// after the completion even if there is a single completion provided.
+	ShellCompDirectiveNoSpace
+
+	// ShellCompDirectiveNoFileComp indicates that the shell should not provide
+	// file completion even when no completion is provided.
+	ShellCompDirectiveNoFileComp
+
+	// ShellCompDirectiveFilterFileExt indicates that the provided completions
+	// should be used as file extension filters.
+	// For flags, using Command.MarkFlagFilename() and Command.MarkPersistentFlagFilename()
+	// is a shortcut to using this directive explicitly.  The BashCompFilenameExt
+	// annotation can also be used to obtain the same behavior for flags.
+	ShellCompDirectiveFilterFileExt
+
+	// ShellCompDirectiveFilterDirs indicates that only directory names should
+	// be provided in file completion.  To request directory names within another
+	// directory, the returned completions should specify the directory within
+	// which to search.  The BashCompSubdirsInDir annotation can be used to
+	// obtain the same behavior but only for flags.
+	ShellCompDirectiveFilterDirs
+
+	// ShellCompDirectiveKeepOrder indicates that the shell should preserve the order
+	// in which the completions are provided
+	ShellCompDirectiveKeepOrder
+
+	// ===========================================================================
+
+	// All directives using iota should be above this one.
+	// For internal use.
+	shellCompDirectiveMaxValue
+
+	// ShellCompDirectiveDefault indicates to let the shell perform its default
+	// behavior after completions have been provided.
+	// This one must be last to avoid messing up the iota count.
+	ShellCompDirectiveDefault ShellCompDirective = 0
+)
+
+// ArgCompletionFunc is a function to do completion for flag value or positional argument.
+type ArgCompletionFunc func(ctx ArgCompletionContext) ([][]string, ShellCompDirective)
+
+// ArgCompletionContext provides essential information to do suggestion
+// for flag value and positional argument completion.
+type ArgCompletionContext interface {
+	context.Context
+
+	Args() []string
+}
+
+type compContextImpl struct {
+	*Context
+	app  *App
+	args []string
+}
+
+func (c compContextImpl) Args() []string {
+	return c.args
+}
+
+// func (c compContextImpl) App() []string {
+// 	return c.app
+// }
+
+// WithArgCompFuncs for struct holding completion functions for use with flags args completion
+func WithArgCompFuncs(funcMap map[string]ArgCompletionFunc) ParseOpt {
+	return ParseOpt{
+		f: func(options *parseOptions) {
+			options.argCompFuncs = funcMap
+		},
+	}
+}
+
+// WithCommandCompFuncs for holding completion functions for use with args completions
+func WithCommandCompFunc(function ArgCompletionFunc) CmdOpt {
+	return CmdOpt{f: func(options *cmdOptions) {
+		options.argCompFunc = function
+	}}
+}
```

Comment:
This option seems unnecessary.

Reply:
Yes, it is to complete positional argument.
You are thinking about unification of arguments passing?

Right now I am passing one function to one command.

Please suggest place to which we would pass this map.

Reply:
See the example `text` above https://github.com/jxskiss/mcli/pull/27#discussion_r1326078956, I passed the map to `Parse`.

Reply:
So the goal is to pass both command positional argument completion and flag completion in same place?

Reply:
Yes, I think this unified API is simpler and more straitforward to use.

Reply:
Yes, but we would get completion after command parsing that way?
I changed the code, but it's not evaluating properly.

Were we would pass it otherwise?

Reply:
Yes, complete positional argument after parsing flags will make it more flex and powerful, i.e. programmer can make use of parsed flags to decide which arguments to suggest.
e.g. for the above example program `say`, if name is "Tom", we suggest "a", "b" as `text`, if name is "John", we suggest "d", "e", "f".
I will take a look into the code.

Reply:
Please do so. I am not pushing my current changes.